Uninstalling extensions fails when extension does not support instances

In https://github.com/exasol/extension-manager/issues/130 we added a check that no instances exists before uninstalling an extensions. This causes now problems for extensions that don't support instances at all, e.g. https://github.com/exasol/kinesis-connector-extension.

Here uninstalling fails with the following error message:

```
2023-10-25 13:38:47.102 [FINE   ] server stderr>ERROR Error processing request: failed to check existing instances: Finding instances not supported 
2023-10-25 13:38:47.104 [FINE   ] server stderr>INFO "DELETE http://localhost:65481/api/v1/extensionmanager/installations/kinesis-connector-extension.js/1.1.1?dbHost=192.168.56.6&dbPort=32784 HTTP/1.1" from 127.0.0.1:65495 - 404 115B in 58.463379ms 
```

Integration tests fail with this exception:

```
org.opentest4j.AssertionFailedError: Unexpected exception thrown: com.exasol.extensionmanager.client.invoker.ApiException: {"code":404,"message":"Finding instances not supported","requestID":"hw3782.clients.exasol.com/0J2E0HKmC8-000002"}

 at com.exasol.cloudetl.kinesis.ExtensionIT.uninstallExtensionWithoutInstallation([ExtensionIT.java:154](vscode-file://vscode-app/Users/chp/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html))
 at java.base/java.util.ArrayList.forEach([ArrayList.java:1541](vscode-file://vscode-app/Users/chp/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html))
 at java.base/java.util.ArrayList.forEach([ArrayList.java:1541](vscode-file://vscode-app/Users/chp/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html))
Caused by: com.exasol.extensionmanager.client.invoker.ApiException: {"code":404,"message":"Finding instances not supported","requestID":"hw3782.clients.exasol.com/0J2E0HKmC8-000002"}

 at com.exasol.extensionmanager.client.invoker.ApiClient.invokeAPI([ApiClient.java:745](vscode-file://vscode-app/Users/chp/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html))
 at com.exasol.extensionmanager.client.api.InstallationApi.uninstallExtension([InstallationApi.java:137](vscode-file://vscode-app/Users/chp/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html))
 at com.exasol.extensionmanager.itest.ExtensionManagerClient.uninstall([ExtensionManagerClient.java:144](vscode-file://vscode-app/Users/chp/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html))
 at com.exasol.extensionmanager.itest.ExtensionManagerClient.uninstall([ExtensionManagerClient.java:140](vscode-file://vscode-app/Users/chp/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html))
 at com.exasol.cloudetl.kinesis.ExtensionIT.lambda$8([ExtensionIT.java:154](vscode-file://vscode-app/Users/chp/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html))
 ... 6 more
```

--- a/.gitignore
+++ b/.gitignore
@@ -78,3 +78,5 @@ vendor/
 
 /registry-upload/node_modules/
 /registry-upload/dist/
+
+**/__debug_bin*

--- a/doc/changes/changes_0.5.5.md
+++ b/doc/changes/changes_0.5.5.md
@@ -1,14 +1,15 @@
-# Extension Manager 0.5.5, released 2023-??-??
+# Extension Manager 0.5.5, released 2023-11-03
 
 Code name: Fix misleading error
 
 ## Summary
 
-This release fixes a misleading error in the integration test framework. When `PreviousVersionManager.prepareBucketFsFile()` is called with an invalid URL it fails with `NoSuchFileException` file exception that hides the actual exception about the invalid URL. The release also improves other error messages in Extension Manager.
+This release fixes a misleading error in the integration test framework. When `PreviousVersionManager.prepareBucketFsFile()` is called with an invalid URL it fails with `NoSuchFileException` file exception that hides the actual exception about the invalid URL. The release also improves other error messages in Extension Manager and fixes a bug that didn't let you uninstall extensions that don't support instances (e.g. because they only require scripts).
 
 ## Bugfixes
 
 * #156: Fixed misleading error in `PreviousVersionManager.prepareBucketFsFile()`
+* #155: Fixed uninstalling extensions that don't support instances
 
 ## Dependency Updates
 

--- a/doc/changes/changes_0.5.5.md
+++ b/doc/changes/changes_0.5.5.md
@@ -1,4 +1,4 @@
-# Extension Manager 0.5.5, released 2023-11-03
+# Extension Manager 0.5.5, released 2023-11-06
 
 Code name: Fix misleading error
 

--- a/pkg/extensionAPI/errorHandlingExtension.go
+++ b/pkg/extensionAPI/errorHandlingExtension.go
@@ -49,6 +49,9 @@ func convertVersions(versions []rawJsExtensionVersion) []JsExtensionVersion {
 }
 
 func (e *JsExtension) GetParameterDefinitions(context *context.ExtensionContext, version string) (definitions []interface{}, errorResult error) {
+	if e.extension.GetParameterDefinitions == nil {
+		return nil, e.unsupportedFunction("getParameterDefinitions")
+	}
 	defer func() {
 		if err := recover(); err != nil {
 			errorResult = e.convertError(fmt.Sprintf("failed to get parameter definitions for extension %q", e.Id), err)

--- a/pkg/extensionAPI/errorHandlingExtension.go
+++ b/pkg/extensionAPI/errorHandlingExtension.go
@@ -130,11 +130,15 @@ func (e *JsExtension) SupportsListInstances(context *context.ExtensionContext, v
 	if err == nil {
 		return true
 	}
+	return !isNotFoundError(err)
+}
+
+func isNotFoundError(err error) bool {
 	apiErr := apiErrors.UnwrapAPIError(err)
 	if apiErr == nil {
-		return true
+		return false
 	}
-	return apiErr.Status != 404
+	return apiErr.Status == 404
 }
 
 func (e *JsExtension) ListInstances(context *context.ExtensionContext, version string) (instances []*JsExtInstance, errorResult error) {

--- a/pkg/extensionAPI/errorHandlingExtension.go
+++ b/pkg/extensionAPI/errorHandlingExtension.go
@@ -122,6 +122,21 @@ func (e *JsExtension) AddInstance(context *context.ExtensionContext, version str
 	return e.extension.AddInstance(context, version, params), nil
 }
 
+func (e *JsExtension) SupportsListInstances(context *context.ExtensionContext, version string) bool {
+	if e.extension.FindInstances == nil {
+		return false
+	}
+	_, err := e.ListInstances(context, version)
+	if err == nil {
+		return true
+	}
+	apiErr := apiErrors.UnwrapAPIError(err)
+	if apiErr == nil {
+		return true
+	}
+	return apiErr.Status != 404
+}
+
 func (e *JsExtension) ListInstances(context *context.ExtensionContext, version string) (instances []*JsExtInstance, errorResult error) {
 	if e.extension.FindInstances == nil {
 		return nil, e.unsupportedFunction("findInstances")

--- a/pkg/extensionAPI/errorHandlingExtension_test.go
+++ b/pkg/extensionAPI/errorHandlingExtension_test.go
@@ -206,6 +206,30 @@ func (suite *ErrorHandlingExtensionSuite) TestUpgradeUnsupported() {
 	suite.Nil(instance)
 }
 
+// ListInstances
+
+func (suite *ErrorHandlingExtensionSuite) TestListInstancesSuccessful() {
+	suite.rawExtension.FindInstances = func(context *context.ExtensionContext, version string) []*JsExtInstance {
+		return []*JsExtInstance{{Id: "inst1", Name: "Inst 1"}, {Id: "inst2", Name: "Inst 2"}}
+	}
+	instances, err := suite.extension.ListInstances(createMockContext(), "version")
+	suite.NoError(err)
+	suite.Equal([]*JsExtInstance{{Id: "inst1", Name: "Inst 1"}, {Id: "inst2", Name: "Inst 2"}}, instances)
+}
+
+func (suite *ErrorHandlingExtensionSuite) TestListInstancesFails() {
+	suite.rawExtension.FindInstances = func(context *context.ExtensionContext, version string) []*JsExtInstance {
+		panic(mockErrorMessage)
+	}
+}
+
+func (suite *ErrorHandlingExtensionSuite) TestListInstancesUnsupported() {
+	suite.rawExtension.FindInstances = nil
+	instances, err := suite.extension.ListInstances(createMockContext(), "version")
+	suite.EqualError(err, `extension "id" does not support operation "findInstances"`)
+	suite.Nil(instances)
+}
+
 // AddInstance
 
 func (suite *ErrorHandlingExtensionSuite) TestAddInstanceSuccessful() {

--- a/pkg/extensionAPI/errorHandlingExtension_test.go
+++ b/pkg/extensionAPI/errorHandlingExtension_test.go
@@ -206,6 +206,27 @@ func (suite *ErrorHandlingExtensionSuite) TestUpgradeUnsupported() {
 	suite.Nil(instance)
 }
 
+// SupportsListInstances
+
+func (suite *ErrorHandlingExtensionSuite) TestSupportsListInstancesUnsupportedMissingMethod() {
+	suite.rawExtension.FindInstances = nil
+	suite.False(suite.extension.SupportsListInstances(createMockContext(), "version"))
+}
+
+func (suite *ErrorHandlingExtensionSuite) TestSupportsListInstancesSupportedReturnsResult() {
+	suite.rawExtension.FindInstances = func(context *context.ExtensionContext, version string) []*JsExtInstance {
+		return []*JsExtInstance{}
+	}
+	suite.True(suite.extension.SupportsListInstances(createMockContext(), "version"))
+}
+
+func (suite *ErrorHandlingExtensionSuite) TestSupportsListInstancesSupportedReturnsGeneralError() {
+	suite.rawExtension.FindInstances = func(context *context.ExtensionContext, version string) []*JsExtInstance {
+		panic(mockErrorMessage)
+	}
+	suite.True(suite.extension.SupportsListInstances(createMockContext(), "version"))
+}
+
 // ListInstances
 
 func (suite *ErrorHandlingExtensionSuite) TestListInstancesSuccessful() {

--- a/pkg/extensionAPI/errorHandlingExtension_test.go
+++ b/pkg/extensionAPI/errorHandlingExtension_test.go
@@ -220,7 +220,7 @@ func (suite *ErrorHandlingExtensionSuite) TestSupportsListInstancesSupportedRetu
 	suite.True(suite.extension.SupportsListInstances(createMockContext(), "version"))
 }
 
-func (suite *ErrorHandlingExtensionSuite) TestSupportsListInstancesSupportedReturnsGeneralError() {
+func (suite *ErrorHandlingExtensionSuite) TestSupportsListInstancesWhenAGeneralErrorOccurs() {
 	suite.rawExtension.FindInstances = func(context *context.ExtensionContext, version string) []*JsExtInstance {
 		panic(mockErrorMessage)
 	}
@@ -242,6 +242,9 @@ func (suite *ErrorHandlingExtensionSuite) TestListInstancesFails() {
 	suite.rawExtension.FindInstances = func(context *context.ExtensionContext, version string) []*JsExtInstance {
 		panic(mockErrorMessage)
 	}
+	instances, err := suite.extension.ListInstances(createMockContext(), "version")
+	suite.EqualError(err, `failed to list instances for extension "id" in version "version": `+mockErrorMessage)
+	suite.Nil(instances)
 }
 
 func (suite *ErrorHandlingExtensionSuite) TestListInstancesUnsupported() {
@@ -255,11 +258,11 @@ func (suite *ErrorHandlingExtensionSuite) TestListInstancesUnsupported() {
 
 func (suite *ErrorHandlingExtensionSuite) TetsGetParameterDefinitionsSuccessful() {
 	suite.rawExtension.GetParameterDefinitions = func(context *context.ExtensionContext, version string) []interface{} {
-		return []interface{}{map[string]interface{}{"id": "param1", "name": "My param", "type": "invalidType"}}
+		return []interface{}{map[string]interface{}{"id": "param1", "name": "My param", "type": "string"}}
 	}
 	instance, err := suite.extension.GetParameterDefinitions(createMockContext(), "version")
 	suite.NoError(err)
-	suite.Equal([]interface{}{map[string]interface{}{"id": "param1", "name": "My param", "type": "invalidType"}}, instance)
+	suite.Equal([]interface{}{map[string]interface{}{"id": "param1", "name": "My param", "type": "string"}}, instance)
 }
 
 func (suite *ErrorHandlingExtensionSuite) TestGetParameterDefinitionsFails() {

--- a/pkg/extensionAPI/errorHandlingExtension_test.go
+++ b/pkg/extensionAPI/errorHandlingExtension_test.go
@@ -230,6 +230,33 @@ func (suite *ErrorHandlingExtensionSuite) TestListInstancesUnsupported() {
 	suite.Nil(instances)
 }
 
+// GetParameterDefinitions
+
+func (suite *ErrorHandlingExtensionSuite) TetsGetParameterDefinitionsSuccessful() {
+	suite.rawExtension.GetParameterDefinitions = func(context *context.ExtensionContext, version string) []interface{} {
+		return []interface{}{map[string]interface{}{"id": "param1", "name": "My param", "type": "invalidType"}}
+	}
+	instance, err := suite.extension.GetParameterDefinitions(createMockContext(), "version")
+	suite.NoError(err)
+	suite.Equal([]interface{}{map[string]interface{}{"id": "param1", "name": "My param", "type": "invalidType"}}, instance)
+}
+
+func (suite *ErrorHandlingExtensionSuite) TestGetParameterDefinitionsFails() {
+	suite.rawExtension.GetParameterDefinitions = func(context *context.ExtensionContext, version string) []interface{} {
+		panic(mockErrorMessage)
+	}
+	instance, err := suite.extension.GetParameterDefinitions(createMockContext(), "version")
+	suite.EqualError(err, `failed to get parameter definitions for extension "id": `+mockErrorMessage)
+	suite.Nil(instance)
+}
+
+func (suite *ErrorHandlingExtensionSuite) TestGetParameterDefinitionsUnsupported() {
+	suite.rawExtension.GetParameterDefinitions = nil
+	instance, err := suite.extension.GetParameterDefinitions(createMockContext(), "version")
+	suite.EqualError(err, `extension "id" does not support operation "getParameterDefinitions"`)
+	suite.Nil(instance)
+}
+
 // AddInstance
 
 func (suite *ErrorHandlingExtensionSuite) TestAddInstanceSuccessful() {

--- a/pkg/extensionController/controller.go
+++ b/pkg/extensionController/controller.go
@@ -234,7 +234,7 @@ func (*controllerImpl) verifyNoInstances(extension *extensionAPI.JsExtension, ex
 	}
 	if len(instances) > 0 {
 		instanceNames := concatInstanceNames(instances)
-		return apiErrors.NewBadRequestErrorF("%d instance(s) exist: %s", len(instances), instanceNames)
+		return apiErrors.NewBadRequestErrorF("cannot uninstall extension because %d instance(s) still exist: %s", len(instances), instanceNames)
 	}
 	return nil
 }

--- a/pkg/extensionController/controller.go
+++ b/pkg/extensionController/controller.go
@@ -217,21 +217,37 @@ func (c *controllerImpl) UninstallExtension(txCtx *transaction.TransactionContex
 		return extensionLoadingFailed(extensionId, err)
 	}
 	extensionCtx := c.createExtensionContext(txCtx)
+	err = c.verifyNoInstances(extension, extensionCtx, extensionVersion)
+	if err != nil {
+		return fmt.Errorf("cannot uninstall extension: %w", err)
+	}
+	return extension.Uninstall(extensionCtx, extensionVersion)
+}
+
+func (*controllerImpl) verifyNoInstances(extension *extensionAPI.JsExtension, extensionCtx *context.ExtensionContext, extensionVersion string) error {
+	if !extension.SupportsListInstances(extensionCtx, extensionVersion) {
+		return nil
+	}
 	instances, err := extension.ListInstances(extensionCtx, extensionVersion)
 	if err != nil {
 		return fmt.Errorf("failed to check existing instances: %w", err)
 	}
 	if len(instances) > 0 {
-		instanceNames := ""
-		for _, inst := range instances {
-			if instanceNames != "" {
-				instanceNames += ", "
-			}
-			instanceNames += inst.Name
-		}
-		return apiErrors.NewBadRequestErrorF("cannot uninstall extension because %d instance(s) still exist: %s", len(instances), instanceNames)
+		instanceNames := concatInstanceNames(instances)
+		return apiErrors.NewBadRequestErrorF("%d instance(s) exist: %s", len(instances), instanceNames)
 	}
-	return extension.Uninstall(extensionCtx, extensionVersion)
+	return nil
+}
+
+func concatInstanceNames(instances []*extensionAPI.JsExtInstance) string {
+	instanceNames := ""
+	for _, inst := range instances {
+		if instanceNames != "" {
+			instanceNames += ", "
+		}
+		instanceNames += inst.Name
+	}
+	return instanceNames
 }
 
 /* [impl -> dsn~upgrade-extension~1]. */

--- a/pkg/extensionController/controller_test.go
+++ b/pkg/extensionController/controller_test.go
@@ -437,7 +437,7 @@ func (suite *ControllerUTestSuite) TestUninstallFailsWhenInstanceExists() {
 	suite.dbMock.ExpectBegin()
 	suite.dbMock.ExpectRollback()
 	err := suite.controller.UninstallExtension(mockContext(), suite.db, EXTENSION_ID, "ver")
-	suite.EqualError(err, "cannot uninstall extension: 2 instance(s) exist: ext-name1, ext-name2")
+	suite.ErrorContains(err, "cannot uninstall extension because 2 instance(s) still exist: ext-name1, ext-name2")
 	apiErr := apiErrors.UnwrapAPIError(err)
 	if apiErr != nil {
 		suite.Equal(400, apiErr.Status)

--- a/pkg/extensionController/controller_test.go
+++ b/pkg/extensionController/controller_test.go
@@ -437,12 +437,26 @@ func (suite *ControllerUTestSuite) TestUninstallFailsWhenInstanceExists() {
 	suite.dbMock.ExpectBegin()
 	suite.dbMock.ExpectRollback()
 	err := suite.controller.UninstallExtension(mockContext(), suite.db, EXTENSION_ID, "ver")
-	suite.EqualError(err, "cannot uninstall extension because 2 instance(s) still exist: ext-name1, ext-name2")
-	if apiErr, ok := apiErrors.AsAPIError(err); ok {
+	suite.EqualError(err, "cannot uninstall extension: 2 instance(s) exist: ext-name1, ext-name2")
+	apiErr := apiErrors.UnwrapAPIError(err)
+	if apiErr != nil {
 		suite.Equal(400, apiErr.Status)
 	} else {
 		suite.Fail(fmt.Sprintf("expected an APIError but got %T: %v", err, err))
 	}
+}
+
+func (suite *ControllerUTestSuite) TestUninstallSucceedsWhenInstancesNotSupported() {
+	integrationTesting.CreateTestExtensionBuilder(suite.T()).
+		WithUninstallFunc("context.sqlClient.execute(`uninstall extension version ${version}`)").
+		WithFindInstancesFunc(`throw new NotFoundError("Finding instances not supported")`).
+		Build().
+		WriteToFile(path.Join(suite.tempExtensionRepo, EXTENSION_ID))
+	suite.dbMock.ExpectBegin()
+	suite.dbMock.ExpectExec("uninstall extension version ver").WillReturnResult(sqlmock.NewResult(0, 0))
+	suite.dbMock.ExpectCommit()
+	err := suite.controller.UninstallExtension(mockContext(), suite.db, EXTENSION_ID, "ver")
+	suite.NoError(err)
 }
 
 func (suite *ControllerUTestSuite) TestUninstallFailsListingInstances() {

--- a/pkg/integrationTesting/extensionForTesting/extensionForTestingTemplate.js
+++ b/pkg/integrationTesting/extensionForTesting/extensionForTestingTemplate.js
@@ -1,6 +1,7 @@
 import {
     BadRequestError,
     InternalServerError,
+    NotFoundError,
     registerExtension
 } from "@exasol/extension-manager-interface";
 
@@ -48,6 +49,10 @@ if (false) {
 if (false) {
     // dummy to keep import
     throw new InternalServerError("dummy");
+}
+if (false) {
+    // dummy to keep import
+    throw new NotFoundError("dummy");
 }
 
 registerExtension(createExtension())

